lxl-web/src/lib/utils/xl.ts

```diff
@@ -0,0 +1,787 @@
+			if (lens[JsonLd.ID]) {
+				lensesById[lens[JsonLd.ID]] = lens;
+			}
+		});
+
+		const flattenedProps = (lens: Lens, hierarchy: string[]): ShowProperties => {
+			if (lens[JsonLd.ID] && hierarchy.includes(lens[JsonLd.ID])) {
+				throw Error(`${Fresnel.extends} inheritance loop: ${hierarchy}`);
+			}
+
+			const superLensId = lens[Fresnel.extends]?.[JsonLd.ID];
+			if (!superLensId) {
+				return lens.showProperties;
+			} else {
+				if (!lensesById[superLensId]) {
+					throw Error(`Super lens not found: ${lens[JsonLd.ID]} ${Fresnel.extends} ${superLensId}`);
+				}
+
+				if (lens['@id']) {
+					hierarchy.push(lens[JsonLd.ID]!);
+				}
+				const superProps = flattenedProps(lensesById[superLensId], hierarchy);
+				let props = lens.showProperties;
+				if (!props.includes(Fresnel.super)) {
+					props = ([Fresnel.super] as ShowProperties).concat(props);
+				}
+				return props.map((p) => (p === Fresnel.super ? superProps : p)).flat();
+			}
+		};
+
+		this.eachLens((lens) => {
+			lens.showProperties = flattenedProps(lens, []);
+			delete lens[Fresnel.extends];
+		});
+	}
+
+	private eachLens(fn: (a: Lens) => void) {
+		const groups = Object.values(this.display.lensGroups).filter(
+			(g) => g[JsonLd.TYPE] === Fresnel.Group
+		); // TODO until "formatters" is moved
+		const lenses = groups.map((g) => Object.values(g.lenses)).flat();
+		for (const lens of lenses) {
+			fn(lens);
+		}
+	}
+}
+
+type Styler = (v: unknown) => unknown;
+
+class Formatter {
+	private readonly DEFAULT_FORMAT: Format = {
+		[JsonLd.TYPE]: Fresnel.Format,
+		[Fresnel.propertyFormat]: {},
+		[Fresnel.valueFormat]: {},
+		[Fresnel.resourceFormat]: {}
+	};
+
+	private readonly stylers: Record<string, Styler> = {
+		'isniGroupDigits()': (v) => {
+			const formatIsni = (isni: unknown) =>
+				typeof isni === 'string' && isni.length === 16
+					? `${isni.slice(0, 4)} ${isni.slice(4, 8)} ${isni.slice(8, 12)} ${isni.slice(12, 16)}`
+					: isni;
+
+			return Array.isArray(v) ? v.map(formatIsni) : formatIsni(v);
+		},
+		'displayType()': (v) => {
+			if (isObject(v) && JsonLd.TYPE in v && Fmt.DISPLAY in v) {
+				// TODO doesn't translate type name
+				(v[Fmt.DISPLAY] as Array<unknown>).unshift({ [JsonLd.VALUE]: v[JsonLd.TYPE] });
+			}
+			return v;
+		}
+	};
+
+	private readonly formatIndex: FormatIndex;
+	private readonly locale: LangCode;
+	private readonly displayUtil: DisplayUtil;
+	private readonly vocabUtil: VocabUtil;
+
+	constructor(
+		displayUtil: DisplayUtil,
+		vocabUtil: VocabUtil,
+		formatIndex: FormatIndex,
+		locale: LangCode
+	) {
+		this.displayUtil = displayUtil;
+		this.vocabUtil = vocabUtil;
+		this.formatIndex = formatIndex;
+		this.locale = locale;
+	}
+
+	displayDecorate(thing: LensedOrdered) {
+		if (!isTypedNode(thing)) {
+			return thing;
+		}
+
+		return this.formatResource(thing, false, false);
+	}
+
+	private formatResource(resource, isFirst: boolean, isLast: boolean) {
+		const className = resource[JsonLd.TYPE];
+		let result = {
```

Comment:
Don't we want to return the `@id` here too, so that we can link to stuff?

Reply:
Haha, yes that was supposed to be the point of this whole machinery :sweat_smile: